pkg/distributor/push.go

```diff
@@ -200,22 +201,27 @@ func handler(
 	})
 }
 
-func calculateRetryAfter(retryAttemptHeader string, baseSeconds int, maxBackoffExponent int) string {
+func calculateRetryAfter(retryAttemptHeader string, minBackoff, maxBackoff time.Duration) string {
+	const jitterFactor = 0.5
+
 	retryAttempt, err := strconv.Atoi(retryAttemptHeader)
 	// If retry-attempt is not valid, set it to default 1
 	if err != nil || retryAttempt < 1 {
 		retryAttempt = 1
 	}
-	if retryAttempt > maxBackoffExponent {
-		retryAttempt = maxBackoffExponent
-	}
-	var minSeconds, maxSeconds int64
-	minSeconds = int64(baseSeconds) << (retryAttempt - 1)
-	maxSeconds = int64(minSeconds) << 1
+	retryAttempt = min(retryAttempt, 63) // Prevent overflow when using retryAttempt to shift
 
-	delaySeconds := minSeconds + rand.Int63n(maxSeconds-minSeconds)
+	delaySeconds := float64(int64(1) << (retryAttempt - 1))
+	delaySeconds = min(maxBackoff.Seconds(), delaySeconds)
+	delaySeconds = max(minBackoff.Seconds(), delaySeconds)
+	if jitterAmount := int64(delaySeconds * jitterFactor); jitterAmount > 0 {
+		delaySeconds += float64(rand.Int63n(jitterAmount))
+	}
+	// Jitter might have pushed the delaySeconds over maxBackoff or minBackoff, so we need to clamp it again.
+	delaySeconds = min(maxBackoff.Seconds(), delaySeconds)
+	delaySeconds = max(minBackoff.Seconds(), delaySeconds)
```

Comment:
I think this can be simplified a bit, to:

```go
	delaySeconds := math.Pow(minBackoff.Seconds(), float64(retryAttempt))
	if jitterAmount := int64(delaySeconds * jitterFactor); jitterAmount > 0 {
		delaySeconds += float64(rand.Int63n(jitterAmount))
	}
	delaySeconds = min(delaySeconds, maxBackoff.Seconds())
```

Reply:
doesn't this use minBackoff as the base? So if minBackoff: 6s, the backoffs will be 6s, 36s, 216s, 1296s,... (with jitter - [3,9], [18,54], [108,324], [648, 1944]...)

aside: I also realized that the intervals were a bit tight and not really with 50% jitter, it was more like 25% in the negative direction, not 50% in either direction. I fixed that in https://github.com/grafana/mimir/pull/8694/commits/c38f5342541c8c75e026d73dc5fb474e1c1e60aa

Reply:
Yes, minBackoff would be the base, and should probably be something like 1s.

Reply:
sounds good; I updated the formula and the docs

Reply:
Cool - with that update I think we can remove a few of the min and max checks, like above?

Reply:
I could remove the first `max(minBackoff, delay)` but I think we need the rest of them. Otherwise, the unit tests fail